archive/oxford_invariants/2021summer/week3_p1.lean

```diff
@@ -0,0 +1,137 @@
+* Base case:
+  * `∑ᵢ₌₀⁰ a₀a₀₊₁/aᵢaᵢ₊₁` is a natural:
+    `∑ᵢ₌₀⁰ a₀a₀₊₁/aᵢaᵢ₊₁ = a₀a₁/a₀a₁ = 1`.
+  * Modulo condition:
+    `a₁ ∣ 0 = a₀ * 1 - a₀`.
+* Induction step:
+  * `∑ᵢ₌₀ⁿ a₀aₙ₊₁/aᵢaᵢ₊₁` is a natural:
+    ```
+    ∑ᵢ₌₀ⁿ⁺¹ a₀aₙ₊₂/aᵢaᵢ₊₁ = ∑ᵢ₌₀ⁿ a₀aₙ₊₂/aᵢaᵢ₊₁ + a₀aₙ₊₂/aₙ₊₁aₙ₊₂
+                        = aₙ₊₂/aₙ₊₁ * ∑ᵢ₌₀ⁿ a₀aₙ₊₁/aᵢaᵢ₊₁ + a₀/aₙ₊₁
+                        = aₙ₊₂/aₙ₊₁ * b + a₀/aₙ₊₁
+                        = (aₙ + aₙ₊₂)/aₙ₊₁ * b - (aₙ * b - a₀)/aₙ₊₁
+    ```
+    which is a natural because `(aₙ + aₙ₊₂)/aₙ₊₁`, `b` and `(aₙ * b - a₀)/aₙ₊₁` (plus an annoying
+    inequality, or the remark that the original sum is positive because its terms are).
+  * Modulo condition:
+    `aₙ₊₁ * ((aₙ + aₙ₊₂)/aₙ₊₁ * b - (aₙ * b - a₀)/aₙ₊₁) - a₀ = aₙ₊₁aₙ₊₂b` is divisible by `aₙ₊₂`.
+-/
+
+open_locale big_operators
+
+theorem week3_p1_aux (n : ℕ) (a : ℕ → ℕ) (a_pos : ∀ i, 0 < a i)
+  (ha : ∀ i < n, a (i + 1) ∣ a i + a (i + 2)) :
+  ∃ b : ℕ, (b : ℚ) = ∑ i in finset.range (n + 1), (a 0 * a (n + 1))/(a i * a (i + 1)) :=
+begin
+  -- Set up the stronger induction hypothesis
+  suffices h : ∃ b : ℕ, (b : ℚ) = ∑ i in finset.range (n + 1), (a 0 * a (n + 1))/(a i * a (i + 1))
+           ∧ a (n + 1) ∣ a n * b - a 0,
+  { obtain ⟨b, hb, -⟩ := h,
+    exact ⟨b, hb⟩ },
+  simp_rw ←@nat.cast_pos ℚ at a_pos,
+  /- Declare the induction
+  `ih` will be the induction hypothesis -/
+  induction n with n ih,
+  /- Base case
+  Claim that the sum equals `1`-/
+  { refine ⟨1, _, _⟩,
+    -- Check that this indeed equals the sum
+    { rw [nat.cast_one, finset.sum_range_one, div_self],
+      exact (mul_pos (a_pos 0) (a_pos 1)).ne' },
+    -- Check the modulo condition
+    { rw [mul_one, nat.sub_self],
+      exact dvd_zero _ } },
+  /- Induction step
+  `b` is the value of the previous sum as a natural, `hb` is the proof that it is indeed the value,
+  and `han` is the divisibility condition -/
+  obtain ⟨b, hb, han⟩ := ih (λ i hi, ha i $ nat.lt_succ_of_lt hi),
+  specialize ha n n.lt_succ_self,
+  have ha₀ : a 0 ≤ a n * b, -- Needing this is an artifact of `ℕ`-substraction.
+  { rw [←@nat.cast_le ℚ, nat.cast_mul, hb, ←div_le_iff' (a_pos _),
+      ←mul_div_mul_right _ _ (a_pos $ n + 1).ne'],
+    suffices h : ∀ i, i ∈ finset.range (n + 1) → 0 ≤ (a 0 : ℚ) * a (n + 1) / (a i * a (i + 1)),
+    { exact finset.single_le_sum h (finset.self_mem_range_succ n) },
+    refine (λ i _, div_nonneg _ _); refine mul_nonneg _ _; exact nat.cast_nonneg _ },
+  -- Claim that the  `(aₙ + aₙ₊₂)/aₙ₊₁ * b - (aₙ * b - a₀)/aₙ₊₁`
+  refine ⟨(a n + a (n + 2))/ a (n + 1) * b - (a n * b - a 0) / a (n + 1), _, _⟩,
+  -- Check that this indeed equals the sum
+  { calc
+      (((a n + a (n + 2)) / a (n + 1) * b - (a n * b - a 0) / a (n + 1) : ℕ) : ℚ)
+        = (a n + a (n + 2)) / a (n + 1) * b - (a n * b - a 0) / a (n + 1) : begin
+          rw [nat.cast_sub, nat.cast_mul, nat.cast_dvd_char_zero ha, nat.cast_add,
+            nat.cast_dvd_char_zero han, nat.cast_sub ha₀, nat.cast_mul],
```

Comment:
Would `norm_cast`/`push_cast` help here?

Reply:
Oh wow! I didn't know that `norm_cast` could use divisibility hypotheses.